# Variáveis e algoritmos

Renato Naville Watanabe

## Algoritmo


Sequência de passos a serem executados para atingir um objetivo ou solução de um problema.

Exemplos de situações onde uma sequência lógica (algoritmo) de passos é necessária:

- Fazer um bolo

- Construir um robô para explorar um local desconhecido

- Trocar uma lâmpada

Um **programa** de computador é a realização (ou implementação) de um **algoritmo** em um computador específico.

É necessário muito cuidado ao fornecer instruções ao computador. Ao fornecer instruções a um computador, ele executará exatamente o que você instruir. Abaixo segue um vídeo do que pode acontecer ao não fornecer instruções precisas para um fim.

In [1]:
from IPython.display import YouTubeVideo
YouTubeVideo('FN2RM-CHkuI')

## Ferramentas de computação científica

Existem diversas ferramentas para utilizadas em cálculos matemáticos avançados.
- Matlab
- Mathematica
- Octave
- Scilab
- R
- **Python**
- Fortran
- Julia

Geralmente contam com bibliotecas de funções matemáticas prontas e recursos avançados.

## Ambientes de desenvolvimento

Existem vários ambientes de programação em Python (conhecidos como IDEs):

* Spyder
* Visual Studio Code
* PyCharm
* Jupyter Notebook
* **Google Colab**
* PyDev
* Atom

## Google Colab


O ambiente de desenvolvimento que será usado no curso é o Google Colab:

<img src="imagens/colab.png" width="500">

De acordo com o que está no site introdutório do Google Colab:

    "O Colaboratory ou "Colab" permite escrever código Python no seu navegador, com:

    * Nenhuma configuração necessária
    * Acesso gratuito a GPUs
    * Compartilhamento fácil
    
    Você pode ser um estudante, um cientista de dados ou um pesquisador de IA, o Colab pode facilitar seu trabalho. "
    
Uma outra vantagem é que não é necessário instalar nada no seu computador.

### Tipos de células

O Google Colab é composto de células. É nas células que são dados os comandos para o Google Colab:

Existem dois tipos de células:

* Texto: é possível escrever textos, fórmulas matemáticas (LaTex) e imagens.
* Código: são dados comandos em Python.

## Dando comandos em Python

Digitando o seguinte comando em uma célula de código e a executando (para executar uma célula basta apertar no triânguo apontando para a direita que está no lado esquerdo da célula ou apertar Shift+Enter):

In [24]:
print(5*2/4 - 0.5)

2.0


é feita a conta $\frac{5\times2}{4}-0,5$ e o seu resultado é mostrado na tela (com o comando print).

Podemos gravar o resultado de uma conta em uma variável, para poder usar o resultado em momento posterior:

In [25]:
x = 5*2/4 - 0.5
print(x)

2.0


## Variável

Uma variável é um endereço da memória de acesso randômico (RAM), representada por um nome 
(rótulo), criado pelo usuário, cujo conteúdo pode se alterar no decorrer do programa.

Uma variável é composta por dois elementos:
    
  - **Identificador**: nome dado pelo programador à variável
  - **Conteúdo**: valor da variável
<img src="imagens/memvar2.png" width="400">

No caso anterior, x é o identificador (nome da variável) e 2 é o conteúdo (valor).

## Variável

Cada vez que mencionarmos o nome de uma variável, estaremos na verdade utilizando o seu conteúdo

In [35]:
y = x + 5
print(y)

8.0


Esta operação define $y$ como sendo uma variável com valor igual ao valor de $x$ **mais** (+) cinco. Ou seja, $y$ tem um valor igual a 7

In [27]:
z = x*y
print(z)

14.0

Neste caso, $z$ é igual à **multiplicação** (\*) dos valores guardados em $x$ e $y$, ou seja, $z$ é igual a 14.

In [28]:
w = z/x
print(w)

7.0

Aqui, $w$ é igual à **divisão** (/) dos valores guardados em $z$ e $x$, ou seja, $w$
é igual a 7.

In [29]:
k = w - x
print(k)

5.0

Aqui, $k$ é igual à **subtração** (-) do valor guardado em $x$ de $w$. Ou seja, $k$
é igual a 5.

In [30]:
m = k**3
print(m)

125.0

Aqui, $m$ é igual ao valor guardado em $k$ **elevado** (\**) ao cubo. Ou seja, $m$ é igual a 125.

In [31]:
n = 10%3
print(n)

1

Neste caso, $n$ tem o valor do **resto** (\%) da divisão entre 10 e 3.

## O símbolo '=' de designação

Em matemática '=' é o símbolo da identidade, mas na programação de computadores '=' é usado para designação. Isso significa que a parte direita da expressão é atribuída à sua parte esquerda. 

Por exemplo, 'x = x + 1' não faz sentido na matemática, mas na programação de computadores sim:



In [32]:
x = x + 1
print(x)

3.0

## Operações aritméticas básicas

Portanto, as operações aritméticas básicas são:

- **multiplicação**: \*
- **divisão**: /
- **adição**: +
- **subtração**: -
- **potência**: \**
- **resto**: \%

## Ordem de operações

Assim como na matemática, em linguagens de programação a multiplicação e a divisão tem precedência sobre soma e subtração.

Assim, na operação:

$$\underbrace{\underbrace{4+\underbrace{3*4}_{12}}_{16}-2}_{14}$$

primeiramente é executada a multiplicação e depois a soma e a subtração na ordem em que elas aparecem.

In [34]:
print(4+3*4-2)

14


A ordem de precedência das operações aritméticas básicas é:

- \** : potência
- \*, /, \% : multiplicação, divisão e resto
- +, - : adição, subtração

No caso de operações com a mesma precedência, prevalece a ordem na qual as operações foram escritas.

Uma maneira de mudar a ordem de execução das operações é utilizando parênteses. As operações dentro de parênteses tem precedência:

In [12]:
print((4+3)*4-2)

26

## Funções matemáticas

Para usar funções mais complexas, é necessário utilizar a biblioteca Numpy. Para poder utilizar esta biblioteca, é necessário importá-la:

In [13]:
import math

Esta biblioteca tem diversas funções que são facilmente utilizadas colocando 'math.' antes da função desejada. Por exemplo, a função logaritmo na base 10 é chamada com: 

In [14]:
print(math.log10(16))

1.2041199826559248

Se você quiser guardar o resultado em uma variável também é possível:

In [15]:
a = math.log10(16)
print(a)

1.2041199826559248

E se quiser calcular o valor da função para uma variável é simples:

In [16]:
b = math.log10(k)
print(b)

0.6989700043360189

## Funções matemáticas

Alguns exemplos de funções são:


**Exponencial**: exp($x$)=$e^x$ em que $e=2,7182...$ 

In [17]:
print(math.exp(-2))

0.1353352832366127

**Raiz quadrada**

In [37]:
s = math.sqrt(25)
print(s)

5.0


**Funções trigonométricas**

Em todas as funções trigonométricas, o ângulo deve estar em radianos.

In [38]:
d = math.sin(np.pi/2)
g = math.cos(np.pi/4)
print(d)
print(g)

1.0
0.7071067811865476


Nos dois últimos casos, para representar o valor de $\pi$, foi utilizado 'np.pi'.

**Função piso e função teto**

A função piso calcula o número inteiro imediatamente menor que o número de entrada da função:

In [20]:
print(math.floor(4.5))

4

A função teto calcula o número inteiro imediatamente maior que o número de entrada da função:

In [21]:
print(math.ceil(4.5))

5

## Comentários

Comentários são expressões que não têm nenhuma função na execução do código, mas servem para auxiliar na compreensão do código.

É muito importante que o código esteja comentado. É muito comum acontecer de um programador ler o seu próprio código dias depois dele ter sido escrito e não entender o que o código faz.

No Python, um comentário é iniciado pelo caracter '#'. Tudo que for escrito após o # não é executado pelo Python

In [22]:
x = x + 1 # Incremento de 1 na variável x
print(x)

4.0

Uma outra maneira de fazer comentários seria com blocos de comentários, iniciados e finalizados com `'''` ou `"""`. Esta é a maneira preferencial para fazer comentários no Python.

## Um código um pouco mais complexo

Em um a célula podem ser dadas várias instruções, uma em cada linha. O código abaixo transforma uma massa em libras para kilogramas.

In [39]:
ml = 10 # massa em libras
mk = ml/2.205 # massa em kg
print('m =',mk,'kg')


m = 4.535147392290249 kg


## Formatando a saída no print

É possível formatar a maneira como resultados numéricos são apresentados na tela. Por exemplo, para mostrar a massa encontrada com apenas três casas decimais, fazemos:

In [42]:
print('m = %.3f kg'%mk)

m = 4.535 kg


## Referências

- Chalco, JM, *Slides de Bases Computacionais da Ciência*, (2014)
- Leite, S, *Slides de Bases Computacionais da Ciência*, (2018)
- [Marietto, MGB et al.; **Bases computacionais da Ciência** (2013)](http://prograd.ufabc.edu.br/images/pdf/bases_computacionais_livro.pdf).
- [Duarte, M, *Python para computação científica* (2017)](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonForScientificComputing_PT-BR.ipynb)
- [NumPy Reference¶](https://docs.scipy.org/doc/numpy/reference/index.html)
- [Duarte, M e Watanabe, R; Python Tutorial (2018)](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonTutorial_PT-BR.ipynb)
